File: backend/src/main/java/com/workflow/engine/form/bizdata/BizDataService.java
package com.workflow.engine.form.bizdata;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.workflow.api.dto.BizDataPageVO;
import com.workflow.api.dto.BizDataQueryRequest;
import com.workflow.api.dto.BizDataVO;
import com.workflow.engine.form.FormDefinitionService;
import com.workflow.engine.form.column.DynamicTableManager;
import com.workflow.engine.tenant.TenantProvider;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * 业务数据服务门面。
 * 按统一顺序路由：覆盖短路 → 守卫检查（update/delete）→ 装饰钩子链 → 通用委托（BizDataSupport）。
 * 通用 SQL 实现与子表读写全部下沉至 {@link BizDataSupport}，本类仅保留路由与编排。
 */
@Service
public class BizDataService {

    /** 覆盖索引的可覆盖操作集合 */
    private static final Set<String> OVERRIDE_OPS = Set.of("create", "update", "delete", "query");

    private final BizDataSupport support;
    private final TenantProvider tenantProvider;
    private final List<FormProcessGuard> guards;
    /** formKey → 钩子列表（按 Spring 注入顺序） */
    private final Map<String, List<BizDataHandler>> handlerIndex;
    /** formKey+操作 → 覆盖 handler（覆盖声明为 true 时的短路径接管，不执行通用实现） */
    private final Map<String, BizDataHandler> coveringIndex;

    /**
     * @param handlers Spring 自动注入所有 BizDataHandler bean（无则空列表）
     * @param guards   Spring 自动注入所有 FormProcessGuard bean（无则空列表）
     */
    public BizDataService(JdbcTemplate jdbcTemplate,
                          DynamicTableManager tableManager,
                          FormDefinitionService formDefService,
                          TenantProvider tenantProvider,
                          ObjectMapper objectMapper,
                          List<BizDataHandler> handlers,
                          List<FormProcessGuard> guards) {
        this.support = new BizDataSupport(jdbcTemplate, tableManager, formDefService, tenantProvider, objectMapper);
        this.tenantProvider = tenantProvider;
        this.guards = guards == null ? List.of() : guards;
        this.handlerIndex = buildHandlerIndex(handlers);
        this.coveringIndex = buildCoveringIndex(handlers);
    }

    private static Map<String, List<BizDataHandler>> buildHandlerIndex(List<BizDataHandler> handlers) {
        Map<String, List<BizDataHandler>> index = new HashMap<>();
        if (handlers == null) {
            return index;
        }
        for (BizDataHandler handler : handlers) {
            if (handler.getFormKey() == null || handler.getFormKey().isBlank()) {
                throw new IllegalStateException("BizDataHandler.getFormKey() 不能为空: " + handler.getClass().getName());
            }
            index.computeIfAbsent(handler.getFormKey(), k -> new ArrayList<>()).add(handler);
        }
        return index;
    }

    /**
     * 构建覆盖索引（key = formKey + "." + 操作 → handler）。
     * 同一 (formKey, op) 出现 2 个及以上有效覆盖声明时抛 IllegalStateException（启动 fail-fast）。
     */
    private static Map<String, BizDataHandler> buildCoveringIndex(List<BizDataHandler> handlers) {
        Map<String, BizDataHandler> index = new HashMap<>();
        if (handlers == null) {
            return index;
        }
        for (BizDataHandler handler : handlers) {
            String formKey = handler.getFormKey();
            for (String op : OVERRIDE_OPS) {
                if (!isOverriding(handler, op)) {
                    continue;
                }
                BizDataHandler previous = index.putIfAbsent(formKey + "." + op, handler);
                if (previous != null) {
                    throw new IllegalStateException("duplicate override declaration: " + formKey + "." + op);
                }
            }
        }
        return index;
    }

    private static boolean isOverriding(BizDataHandler handler, String op) {
        return switch (op) {
            case "create" -> handler.overridesCreate();
            case "update" -> handler.overridesUpdate();
            case "delete" -> handler.overridesDelete();
            case "query" -> handler.overridesQuery();
            default -> false;
        };
    }

    private List<BizDataHandler> handlersOf(String formKey) {
        return handlerIndex.getOrDefault(formKey, List.of());
    }

    // ==================== 主表 CRUD 入口 ====================

    /**
     * 新增业务数据（覆盖短路 → 装饰钩子链 → 通用委托）。
     */
    @Transactional
    public BizDataVO create(String formKey, Map<String, Object> data) {
        BizDataHandler covering = coveringIndex.get(formKey + ".create");
        if (covering != null) {
            return covering.create(data);
        }
        support.loadContext(formKey);
        for (BizDataHandler handler : handlersOf(formKey)) {
            handler.beforeCreate(data);
        }
        BizDataVO created = support.createGeneric(formKey, data);
        for (BizDataHandler handler : handlersOf(formKey)) {
            handler.afterCreate(created);
        }
        return created;
    }

    /**
     * 分页查询业务数据（覆盖短路 → 通用委托）。
     */
    public BizDataPageVO query(String formKey, BizDataQueryRequest req) {
        BizDataHandler covering = coveringIndex.get(formKey + ".query");
        if (covering != null) {
            return covering.query(req);
        }
        return support.queryGeneric(formKey, req);
    }

    /**
     * 查询单条业务数据。
     */
    public BizDataVO getById(String formKey, String id) {
        BizDataContext ctx = support.loadContext(formKey);
        return support.findById(ctx.tableName(), tenantProvider.getTenantId(), ctx, id);
    }

    /**
     * 更新业务数据（乐观锁；覆盖短路 → 守卫检查 → 装饰钩子链 → 通用委托）。
     */
    @Transactional
    public BizDataVO update(String formKey, String id, Map<String, Object> data, Integer version) {
        BizDataHandler covering = coveringIndex.get(formKey + ".update");
        if (covering != null) {
            return covering.update(id, data, version);
        }
        BizDataContext ctx = support.loadContext(formKey);
        BizDataVO existing = support.findById(ctx.tableName(), tenantProvider.getTenantId(), ctx, id);
        for (BizDataHandler handler : handlersOf(formKey)) {
            handler.beforeUpdate(data, existing);
        }
        BizDataVO updated = support.updateGeneric(formKey, id, data, version);
        return updated;
    }

    /**
     * 删除业务数据（租户范围限定；覆盖短路 → 守卫检查 → 装饰钩子链 → 通用委托）。
     */
    @Transactional
    public void delete(String formKey, String id) {
        BizDataHandler covering = coveringIndex.get(formKey + ".delete");
        if (covering != null) {
            covering.delete(id);
            return;
        }
        BizDataContext ctx = support.loadContext(formKey);
        BizDataVO existing = support.findById(ctx.tableName(), tenantProvider.getTenantId(), ctx, id);
        for (BizDataHandler handler : handlersOf(formKey)) {
            handler.beforeDelete(existing);
        }
        support.deleteGeneric(formKey, id);
    }

    // ==================== 独立子表行 CRUD ====================

    /**
     * 分页查询独立子表行（sort_no 升序）。
     */
    public List<Map<String, Object>> listSubRows(String formKey, String id, String field) {
        return support.listSubRows(formKey, id, field);
    }

    /**
     * 新增独立子表行（追加到末尾，sort_no 续接）。
     */
    @Transactional
    public Map<String, Object> addSubRow(String formKey, String id, String field, Map<String, Object> data) {
        return support.addSubRow(formKey, id, field, data);
    }

    /**
     * 更新独立子表行（乐观锁：须携带当前 version）。
     */
    @Transactional
    public Map<String, Object> updateSubRow(String formKey, String id, String field, String rowId,
                                            Map<String, Object> data, Integer version) {
        return support.updateSubRow(formKey, id, field, rowId, data, version);
    }

    /**
     * 删除独立子表行。
     */
    @Transactional
    public void deleteSubRow(String formKey, String id, String field, String rowId) {
        support.deleteSubRow(formKey, id, field, rowId);
    }

    // ==================== 引用解析 ====================

    /**
     * 按表单 key 批量解析显示文本（resolve API 入口）。
     */
    public Map<String, String> resolveByFormKey(String formKey, List<String> ids, String displayField) {
        return support.resolveByFormKey(formKey, ids, displayField);
    }

    /**
     * 批量解析被引用记录的显示文本（id → displayField 值）。
     */
    public Map<String, String> resolveDisplayTexts(String sourceFormKey, List<String> ids, String displayField) {
        return support.resolveDisplayTexts(sourceFormKey, ids, displayField);
    }

    /**
     * 统计各业务表单被 dataPicker 引用的情况（引用感知）。
     */
    public Map<String, Map<String, Object>> countReferencedBy() {
        return support.countReferencedBy();
    }
}

File: backend-node/src/engine/form/bizdata/biz-data.service.ts
import { Injectable } from '@nestjs/common'
import { BusinessException } from '../../../common/exception/business-exception'
import { getTenantId } from '../../../framework/tenant/tenant-context'
import { BizDataSupport, type BizDataQueryRequest } from './biz-data-support'
import type { FormQueryConfig } from './form-query-config'
import type { JoinConfig } from './join-sql-generator'
import type { BizDataPageVO, BizDataVO } from '../../../common/domain/biz-data'

/**
 * 业务数据门面（对齐 Java `com.workflow.engine.form.bizdata.BizDataService`）。
 *
 * Java 侧这一层还有「业务定制 handler 覆盖」与「流程守卫」两套扩展点
 * （`coveringIndex` / `guards`）。当前系统里注册的 handler 只覆盖
 * `emp_profile` 与 `leave_bill` 两个示例表单，与本次迁移的表单无交集，
 * 因此没有实现 —— 但**必须留痕**：一旦将来有表单需要定制，
 * 这里是唯一的接入点，不要绕过它直接调 `BizDataSupport`。
 */
@Injectable()
export class BizDataService {
  constructor(private readonly support: BizDataSupport) {}

  /** 已发布业务表单的列映射（供数据源适配器组装 metadata）。 */
  async loadColumns(formKey: string) {
    return this.support.loadColumns(formKey)
  }

  /** 分页查询业务数据。 */
  async query(formKey: string, req: BizDataQueryRequest): Promise<BizDataPageVO> {
    return this.support.queryGeneric(formKey, req)
  }

  /**
   * config 模式分页查询（声明式 JOIN + 虚拟列，对齐 Java `BizDataService.queryJoin`）。
   *
   * Java 在这一层还会先查业务定制 handler（`coveringIndex`）；当前系统注册的 handler
   * 只覆盖 `emp_profile` / `leave_bill` 两个示例表单，与本次迁移的表单无交集，
   * 因此与其余方法一致地**不实现** handler 覆盖（见类注释留痕）。
   */
  async queryJoin(
    formKey: string,
    req: BizDataQueryRequest,
    joins: JoinConfig[],
  ): Promise<BizDataPageVO> {
    return this.support.queryJoinConfig(formKey, req, joins)
  }

  /** sql 模式分页查询（走业务定制 handler，对齐 Java `BizDataService.querySql`）。 */
  async querySql(
    formKey: string,
    req: BizDataQueryRequest,
    config: FormQueryConfig,
  ): Promise<BizDataPageVO> {
    return this.support.querySqlTemplate(formKey, req, config)
  }

  /**
   * sql 模式分页查询（**绕过**业务定制 handler）。
   *
   * SQL 数据源执行的是管理员显式 SQL，不能被绑定表单的业务定制劫持
   * （对齐 Java `BizDataService.querySqlRaw`）。
   */
  async querySqlRaw(
    formKey: string | null,
    req: BizDataQueryRequest,
    config: FormQueryConfig,
  ): Promise<BizDataPageVO> {
    return this.support.querySqlTemplate(formKey, req, config)
  }

  /** 查询单条业务数据。 */
  async getById(formKey: string, id: string): Promise<BizDataVO> {
    const ctx = await this.support.loadContext(formKey)
    return this.support.findById(ctx, getTenantId(), id)
  }

  /** 按表单 key 批量解析显示文本。 */
  async resolveByFormKey(
    formKey: string,
    ids: string[],
    displayField: string | null,
  ): Promise<Record<string, string>> {
    return this.support.resolveByFormKey(formKey, ids, displayField)
  }

  /** 批量解析被引用记录的显示文本。 */
  async resolveDisplayTexts(
    sourceFormKey: string,
    ids: string[],
    displayField: string,
  ): Promise<Record<string, string>> {
    return this.support.resolveDisplayTexts(sourceFormKey, ids, displayField)
  }

  /** 统计各业务表单被 dataPicker 引用的情况。 */
  async countReferencedBy() {
    return this.support.countReferencedBy()
  }

  /**
   * 分页查询独立子表行。
   *
   * 对齐 Java `BizDataService.listSubRows`。**校验顺序不可换**：
   * `loadContext` → `requireSubTable`（404）→ `requireMainRow`（404）→ 读子表。
   * 顺序错了，同一个请求会从「字段不存在」变成「主行不存在」，错误消息就分叉了。
   */
  async listSubRows(
    formKey: string,
    id: string,
    field: string,
  ): Promise<Array<Record<string, unknown>>> {
    const ctx = await this.support.loadContext(formKey)
    const def = ctx.subTables.get(field)
    if (def === undefined) {
      throw new BusinessException(404, `子表字段不存在: ${field}`)
    }
    // 主表行不存在也返回 404（与 Java 的 requireMainRow 一致）
    await this.support.findById(ctx, getTenantId(), id)
    return this.support.readSubTableRows(def, id)
  }

  // ==================== 写路径（原为 notMigratedWrite 占位） ====================

  /** 新增业务数据（`BizDataService.create`）。 */
  async create(formKey: string, data: Record<string, unknown> | null): Promise<BizDataVO> {
    return this.support.createGeneric(formKey, data)
  }

  /** 更新业务数据（乐观锁；`version` 来自请求体）。 */
  async update(
    formKey: string,
    id: string,
    data: Record<string, unknown> | null,
    version: number | null,
  ): Promise<BizDataVO> {
    return this.support.updateGeneric(formKey, id, data, version)
  }

  /** 删除业务数据（级联删除子表行）。 */
  async remove(formKey: string, id: string): Promise<void> {
    await this.support.deleteGeneric(formKey, id)
  }

  /** 新增独立子表行。 */
  async addSubRow(
    formKey: string,
    id: string,
    field: string,
    data: Record<string, unknown> | null,
  ): Promise<Record<string, unknown>> {
    return this.support.addSubRow(formKey, id, field, data)
  }

  /** 更新独立子表行（乐观锁；`version` 来自请求体）。 */
  async updateSubRow(
    formKey: string,
    id: string,
    field: string,
    rowId: string,
    data: Record<string, unknown> | null,
    version: number | null,
  ): Promise<Record<string, unknown>> {
    return this.support.updateSubRow(formKey, id, field, rowId, data, version)
  }

  /** 删除独立子表行。 */
  async deleteSubRow(
    formKey: string,
    id: string,
    field: string,
    rowId: string,
  ): Promise<void> {
    await this.support.deleteSubRow(formKey, id, field, rowId)
  }
}
